README
NAME
    Drupal::Credentials - Access credentials (user, pass etc.) from Drupal's
    sites directory (settings.php)

VERSION
    Version 0.02

SYNOPSIS
    Access info from Drupal's $db_uri contained in settings.php

        use Drupal::Credentials;

    #new Drupal::Credentials::Symlinks='follow';#default is dontfollow

        my $credentials = Drupal::Credentials->new($site_dir);

            foreach my $site_id ($credentials->list) {
                    my $dbstring=$credentials->get_dbstring($site_id);
                    my $proto=$credentials->get_scheme($site_id);
                    my $passw=$credentials->get_pass($site_id);
                    my $db=$credentials->get_database($site_id);
                    my $host=$credentials->get_host($site_id);
                    my $user=$credentials->get_user($site_id);
    #new:
                    my $dir=$credentials->get_install_dir ($site_id);
            }
            my $sites_dir=$credentials->get_sites_dir;

METHODS
  my $credentials = Drupal::Credentials->new($site_dir);
    $site_dir is the full path to the drupal/sites directory.

    Returns the Drupal::Credentials object on success and nothing on
    failure.

  $credentials->_parse_sites_dir
    Reads the directory specified as sites_dir during new, and looks for
    subdirectories which a setting.php file in them. Settings.php file is
    searched for $db_url from which the credentials are extracted.

    TODO: Currently, always returns 1. Should be able to fail

  my $host=$credentials->get_host($site_id);
    Return host from credentials.

  my $pw=$credentials->get_pass($site_id);
    Return pass from credentials for site with $site_id.

  my $proto=$credentials->get_scheme($site_id);
    Return scheme from credentials for site with $site_id.

  my $un=$credentials->get_user($site_id);
    Return user from credentials for site with $site_id.

  my $db=$credentials->get_database($site_id);
    Return database from credentials for site with $site_id.

  my $dbstring=$credentials->get_dbstring($site_id);
    Return string from settings.php's $db_uri for site with $site_id. This
    one will not be unescaped, if it uses hex encoding (e.g. %2a for :). I
    suspect that this method is not of real use except for debugging.

  my $credentials->list;
Return site ids in an array. A site id is the name of the directory in sites
directory, e.g.
	exampledomain.com
    (In code site id is sometimes referred to as site name.)

  my $sites_dir=$credentials->get_install_dir;
    Return path for the installation (in Drupal's site directory). At the
    moment, it can be relative or absolute path.

    Todo: Do i need an absolute path?

  my $sites_dir=$credentials->get_sites_dir;
    Return Drupal's path for site as specified during new.

SEE ALSO
    http://php.net/manual/en/function.parse-url.php Drupal's settings.php

AUTHOR
    Maurice Mengel, "<mauricemengel at gmail.com>"

BUGS
    Please report any bugs or feature requests to "bug-drupal-credentials at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Drupal-Credentials>. I
    will be notified, and then you'll automatically be notified of progress
    on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Drupal::Credentials

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Drupal-Credentials>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Drupal-Credentials>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Drupal-Credentials>

    *   Search CPAN

        <http://search.cpan.org/dist/Drupal-Credentials/>

INSTALLATION
    To install this module, run the following commands:

            perl Makefile.PL
            make
            make test
            make install

LICENSE AND COPYRIGHT
    Copyright 2010 Maurice Mengel.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Drupal-Credentials

see http://github.com/mokko/Drupal-Credentials/commits/master/

test.pl
#!/usr/bin/perl

use strict;
use warnings;
use lib 'lib';
use Drupal::Credentials;
use Data::Dumper;
$Data::Dumper::Indent = 1;
#$Drupal::Credentials::Symlinks='follow';
my $credentials = Drupal::Credentials->new('t/sites');


print "in test.pl $Drupal::Credentials::Symlinks \n";
foreach ( $credentials->list ) {
	print "TT$_\n";
}

print "d" . Dumper $credentials;


t/sites/default/settings.php
<?php
// $Id: default.settings.php,v 1.8.2.1 2008/08/13 06:52:36 dries Exp $

/**
 * @file
 * Drupal site-specific configuration file.
 *
 * IMPORTANT NOTE:
 * This file may have been set to read-only by the Drupal installation
 * program. If you make changes to this file, be sure to protect it again
 * after making your modifications. Failure to remove write permissions
 * to this file is a security risk.
 *
 * The configuration file to be loaded is based upon the rules below.
 *
 * The configuration directory will be discovered by stripping the
 * website's hostname from left to right and pathname from right to
 * left. The first configuration file found will be used and any
 * others will be ignored. If no other configuration file is found
 * then the default configuration file at 'sites/default' will be used.
 *
 * For example, for a fictitious site installed at
 * http://www.drupal.org/mysite/test/, the 'settings.php'
 * is searched in the following directories:
 *
 *  1. sites/www.drupal.org.mysite.test
 *  2. sites/drupal.org.mysite.test
 *  3. sites/org.mysite.test
 *
 *  4. sites/www.drupal.org.mysite
 *  5. sites/drupal.org.mysite
 *  6. sites/org.mysite
 *
 *  7. sites/www.drupal.org
 *  8. sites/drupal.org
 *  9. sites/org
 *
 * 10. sites/default
 *
 * If you are installing on a non-standard port number, prefix the
 * hostname with that number. For example,
 * http://www.drupal.org:8080/mysite/test/ could be loaded from
 * sites/8080.www.drupal.org.mysite.test/.
 */

/**
 * Database settings:
 *
 * Note that the $db_url variable gets parsed using PHP's built-in
 * URL parser (i.e. using the "parse_url()" function) so make sure
 * not to confuse the parser. If your username, password
 * or database name contain characters used to delineate
 * $db_url parts, you can escape them via URI hex encodings:
 *
 *   : = %3a   / = %2f   @ = %40
 *   + = %2b   ( = %28   ) = %29
 *   ? = %3f   = = %3d   & = %26
 *
 * To specify multiple connections to be used in your site (i.e. for
 * complex custom modules) you can also specify an associative array
 * of $db_url variables with the 'default' element used until otherwise
 * requested.
 *
 * You can optionally set prefixes for some or all database table names
 * by using the $db_prefix setting. If a prefix is specified, the table
 * name will be prepended with its value. Be sure to use valid database
 * characters only, usually alphanumeric and underscore. If no prefixes
 * are desired, leave it as an empty string ''.
 *
 * To have all database names prefixed, set $db_prefix as a string:
 *
 *   $db_prefix = 'main_';
 *
 * To provide prefixes for specific tables, set $db_prefix as an array.
 * The array's keys are the table names and the values are the prefixes.
 * The 'default' element holds the prefix for any tables not specified
 * elsewhere in the array. Example:
 *
 *   $db_prefix = array(
 *     'default'   => 'main_',
 *     'users'     => 'shared_',
 *     'sessions'  => 'shared_',
 *     'role'      => 'shared_',
 *     'authmap'   => 'shared_',
 *   );
 *
 * Database URL format:
 *   $db_url = 'mysql://username:password@localhost/databasename';
 *   $db_url = 'mysqli://username:password@localhost/databasename';
 *   $db_url = 'pgsql://username:password@localhost/databasename';
 */
$db_url = 'mysqli://mimose:hornibostelchen@localhost/mimose';
$db_prefix = 'main_';

/**
 * Access control for update.php script
 *
 * If you are updating your Drupal installation using the update.php script
 * being not logged in as administrator, you will need to modify the access
 * check statement below. Change the FALSE to a TRUE to disable the access
 * check. After finishing the upgrade, be sure to open this file again
 * and change the TRUE back to a FALSE!
 */
$update_free_access = FALSE;

/**
 * Base URL (optional).
 *
 * If you are experiencing issues with different site domains,
 * uncomment the Base URL statement below (remove the leading hash sign)
 * and fill in the URL to your Drupal installation.
 *
 * You might also want to force users to use a given domain.
 * See the .htaccess file for more information.
 *
 * Examples:
 *   $base_url = 'http://www.example.com';
 *   $base_url = 'http://www.example.com:8888';
 *   $base_url = 'http://www.example.com/drupal';
 *   $base_url = 'https://www.example.com:8888/drupal';
 *
 * It is not allowed to have a trailing slash; Drupal will add it
 * for you.
 */
#$base_url = 'http://www.mimo-project.eu';  // NO trailing slash!

/**
 * PHP settings:
 *
 * To see what PHP settings are possible, including whether they can
 * be set at runtime (ie., when ini_set() occurs), read the PHP
 * documentation at http://www.php.net/manual/en/ini.php#ini.list
 * and take a look at the .htaccess file to see which non-runtime
 * settings are used there. Settings defined here should not be
 * duplicated there so as to avoid conflict issues.
 */
ini_set('arg_separator.output',     '&amp;');
ini_set('magic_quotes_runtime',     0);
ini_set('magic_quotes_sybase',      0);
ini_set('session.cache_expire',     200000);
ini_set('session.cache_limiter',    'none');
ini_set('session.cookie_lifetime',  2000000);
ini_set('session.gc_maxlifetime',   200000);
ini_set('session.save_handler',     'user');
ini_set('session.use_only_cookies', 1);
ini_set('session.use_trans_sid',    0);
ini_set('url_rewriter.tags',        '');

/**
 * Drupal automatically generates a unique session cookie name for each site
 * based on on its full domain name. If you have multiple domains pointing at
 * the same Drupal site, you can either redirect them all to a single domain
 * (see comment in .htaccess), or uncomment the line below and specify their
 * shared base domain. Doing so assures that users remain logged in as they
 * cross between your various domains.
 */
# $cookie_domain = 'example.com';

/**
 * Variable overrides:
 *
 * To override specific entries in the 'variable' table for this site,
 * set them here. You usually don't need to use this feature. This is
 * useful in a configuration file for a vhost or directory, rather than
 * the default settings.php. Any configuration setting from the 'variable'
 * table can be given a new value. Note that any values you provide in
 * these variable overrides will not be modifiable from the Drupal
 * administration interface.
 *
 * Remove the leading hash signs to enable.
 */
# $conf = array(
#   'site_name' => 'My Drupal site',
#   'theme_default' => 'minnelli',
#   'anonymous' => 'Visitor',
/**
 * A custom theme can be set for the off-line page. This applies when the site
 * is explicitly set to off-line mode through the administration page or when
 * the database is inactive due to an error. It can be set through the
 * 'maintenance_theme' key. The template file should also be copied into the
 * theme. It is located inside 'modules/system/maintenance-page.tpl.php'.
 * Note: This setting does not apply to installation and update pages.
 */
#   'maintenance_theme' => 'minnelli',
/**
 * reverse_proxy accepts a boolean value.
 *
 * Enable this setting to determine the correct IP address of the remote
 * client by examining information stored in the X-Forwarded-For headers.
 * X-Forwarded-For headers are a standard mechanism for identifying client
 * systems connecting through a reverse proxy server, such as Squid or
 * Pound. Reverse proxy servers are often used to enhance the performance
 * of heavily visited sites and may also provide other site caching,
 * security or encryption benefits. If this Drupal installation operates
 * behind a reverse proxy, this setting should be enabled so that correct
 * IP address information is captured in Drupal's session management,
 * logging, statistics and access management systems; if you are unsure
 * about this setting, do not have a reverse proxy, or Drupal operates in
 * a shared hosting environment, this setting should be set to disabled.
 */
#   'reverse_proxy' => TRUE,
/**
 * reverse_proxy accepts an array of IP addresses.
 *
 * Each element of this array is the IP address of any of your reverse
 * proxies. Filling this array Drupal will trust the information stored
 * in the X-Forwarded-For headers only if Remote IP address is one of
 * these, that is the request reaches the web server from one of your
 * reverse proxies. Otherwise, the client could directly connect to
 * your web server spoofing the X-Forwarded-For headers.
 */
#   'reverse_proxy_addresses' => array('a.b.c.d', ...),
# );

/**
 * String overrides:
 *
 * To override specific strings on your site with or without enabling locale
 * module, add an entry to this list. This functionality allows you to change
 * a small number of your site's default English language interface strings.
 *
 * Remove the leading hash signs to enable.
 */
# $conf['locale_custom_strings_en'] = array(
#   'forum'      => 'Discussion board',
#   '@count min' => '@count minutes',
# );

/**
 * Multilingual settings
 */
$conf['i18n_variables']= array(
// Site name, slogan ...
//'site_name',
'site_slogan',
'site_mission',
'site_footer',
//'anonymous',
//Different front page for each language
'site_frontpage',
'menu_primary_links_source',
'menu_secondary_links_source',
'contact_form_information',
'custom_search_box_label', 'custom_search_box_default', 'custom_search_box_submit_text'
);




t/sites/test.mimo-project.eu/settings.php
<?php
// $Id: default.settings.php,v 1.8 2007/12/20 09:35:10 goba Exp $

/**
 * @file
 * Drupal site-specific configuration file.
 *
 * IMPORTANT NOTE:
 * This file may have been set to read-only by the Drupal installation
 * program. If you make changes to this file, be sure to protect it again
 * after making your modifications. Failure to remove write permissions
 * to this file is a security risk.
 *
 * The configuration file to be loaded is based upon the rules below.
 *
 * The configuration directory will be discovered by stripping the
 * website's hostname from left to right and pathname from right to
 * left. The first configuration file found will be used and any
 * others will be ignored. If no other configuration file is found
 * then the default configuration file at 'sites/default' will be used.
 *
 * For example, for a fictitious site installed at
 * http://www.drupal.org/mysite/test/, the 'settings.php'
 * is searched in the following directories:
 *
 *  1. sites/www.drupal.org.mysite.test
 *  2. sites/drupal.org.mysite.test
 *  3. sites/org.mysite.test
 *
 *  4. sites/www.drupal.org.mysite
 *  5. sites/drupal.org.mysite
 *  6. sites/org.mysite
 *
 *  7. sites/www.drupal.org
 *  8. sites/drupal.org
 *  9. sites/org
 *
 * 10. sites/default
 *
 * If you are installing on a non-standard port number, prefix the
 * hostname with that number. For example,
 * http://www.drupal.org:8080/mysite/test/ could be loaded from
 * sites/8080.www.drupal.org.mysite.test/.
 */

/**
 * Database settings:
 *
 * Note that the $db_url variable gets parsed using PHP's built-in
 * URL parser (i.e. using the "parse_url()" function) so make sure
 * not to confuse the parser. If your username, password
 * or database name contain characters used to delineate
 * $db_url parts, you can escape them via URI hex encodings:
 *
 *   : = %3a   / = %2f   @ = %40
 *   + = %2b   ( = %28   ) = %29
 *   ? = %3f   = = %3d   & = %26
 *
 * To specify multiple connections to be used in your site (i.e. for
 * complex custom modules) you can also specify an associative array
 * of $db_url variables with the 'default' element used until otherwise
 * requested.
 *
 * You can optionally set prefixes for some or all database table names
 * by using the $db_prefix setting. If a prefix is specified, the table
 * name will be prepended with its value. Be sure to use valid database
 * characters only, usually alphanumeric and underscore. If no prefixes
 * are desired, leave it as an empty string ''.
 *
 * To have all database names prefixed, set $db_prefix as a string:
 *
 * $db_prefix = 'main_';
 *
 * To provide prefixes for specific tables, set $db_prefix as an array.
 * The array's keys are the table names and the values are the prefixes.
 * The 'default' element holds the prefix for any tables not specified
 * elsewhere in the array. Example:
 *
 *   $db_prefix = array(
 *     'default'   => 'main_',
 *     'users'     => 'shared_',
 *     'sessions'  => 'shared_',
 *     'role'      => 'shared_',
 *     'authmap'   => 'shared_',
 *   );
 *
 * Database URL format:
 *   $db_url = 'mysql://username:password@localhost/databasename';
 *   $db_url = 'mysqli://username:password@localhost/databasename';
 *   $db_url = 'pgsql://username:password@localhost/databasename';
 */
$db_url = 'mysql://mimo:hornbostel_2009@localhost/mimo-project';
$db_prefix = 'test_';

/**
 * Access control for update.php script
 *
 * If you are updating your Drupal installation using the update.php script
 * being not logged in as administrator, you will need to modify the access
 * check statement below. Change the FALSE to a TRUE to disable the access
 * check. After finishing the upgrade, be sure to open this file again
 * and change the TRUE back to a FALSE!
 */
$update_free_access = FALSE;

/**
 * Base URL (optional).
 *
 * If you are experiencing issues with different site domains,
 * uncomment the Base URL statement below (remove the leading hash sign)
 * and fill in the URL to your Drupal installation.
 *
 * You might also want to force users to use a given domain.
 * See the .htaccess file for more information.
 *
 * Examples:
 *   $base_url = 'http://www.example.com';
 *   $base_url = 'http://www.example.com:8888';
 *   $base_url = 'http://www.example.com/drupal';
 *   $base_url = 'https://www.example.com:8888/drupal';
 *
 * It is not allowed to have a trailing slash; Drupal will add it
 * for you.
 */
# $base_url = 'http://www.example.com';  // NO trailing slash!
# maurice
$base_url='http://test.mimo-project.eu';

/**
 * PHP settings:
 *
 * To see what PHP settings are possible, including whether they can
 * be set at runtime (ie., when ini_set() occurs), read the PHP
 * documentation at http://www.php.net/manual/en/ini.php#ini.list
 * and take a look at the .htaccess file to see which non-runtime
 * settings are used there. Settings defined here should not be
 * duplicated there so as to avoid conflict issues.
 */
ini_set('arg_separator.output',     '&amp;');
ini_set('magic_quotes_runtime',     0);
ini_set('magic_quotes_sybase',      0);
ini_set('session.cache_expire',     200000);
ini_set('session.cache_limiter',    'none');
ini_set('session.cookie_lifetime',  2000000);
ini_set('session.gc_maxlifetime',   200000);
ini_set('session.save_handler',     'user');
ini_set('session.use_only_cookies', 1);
ini_set('session.use_trans_sid',    0);
ini_set('url_rewriter.tags',        '');

/**
 * Drupal automatically generates a unique session cookie name for each site
 * based on on its full domain name. If you have multiple domains pointing at
 * the same Drupal site, you can either redirect them all to a single domain
 * (see comment in .htaccess), or uncomment the line below and specify their
 * shared base domain. Doing so assures that users remain logged in as they
 * cross between your various domains.
 */
#$cookie_domain = 'www.musikethnologie.org';

/**
 * Variable overrides:
 *
 * To override specific entries in the 'variable' table for this site,
 * set them here. You usually don't need to use this feature. This is
 * useful in a configuration file for a vhost or directory, rather than
 * the default settings.php. Any configuration setting from the 'variable'
 * table can be given a new value. Note that any values you provide in
 * these variable overrides will not be modifiable from the Drupal
 * administration interface.
 *
 * Remove the leading hash signs to enable.
 */
# $conf = array(
#   'site_name' => 'My Drupal site',
#   'theme_default' => 'minnelli',
#   'anonymous' => 'Visitor',
/**
 * A custom theme can be set for the off-line page. This applies when the site
 * is explicitly set to off-line mode through the administration page or when
 * the database is inactive due to an error. It can be set through the
 * 'maintenance_theme' key. The template file should also be copied into the
 * theme. It is located inside 'modules/system/maintenance-page.tpl.php'.
 * Note: This setting does not apply to installation and update pages.
 */
#   'maintenance_theme' => 'minnelli',
/**
 * reverse_proxy accepts a boolean value.
 *
 * Enable this setting to determine the correct IP address of the remote
 * client by examining information stored in the X-Forwarded-For headers.
 * X-Forwarded-For headers are a standard mechanism for identifying client
 * systems connecting through a reverse proxy server, such as Squid or
 * Pound. Reverse proxy servers are often used to enhance the performance
 * of heavily visited sites and may also provide other site caching,
 * security or encryption benefits. If this Drupal installation operates
 * behind a reverse proxy, this setting should be enabled so that correct
 * IP address information is captured in Drupal's session management,
 * logging, statistics and access management systems; if you are unsure
 * about this setting, do not have a reverse proxy, or Drupal operates in
 * a shared hosting environment, this setting should be set to disabled.
 */
#   'reverse_proxy' => TRUE,
/**
 * reverse_proxy accepts an array of IP addresses.
 *
 * Each element of this array is the IP address of any of your reverse
 * proxies. Filling this array Drupal will trust the information stored
 * in the X-Forwarded-For headers only if Remote IP address is one of
 * these, that is the request reaches the web server from one of your
 * reverse proxies. Otherwise, the client could directly connect to
 * your web server spoofing the X-Forwarded-For headers.
 */
#   'reverse_proxy_addresses' => array('a.b.c.d', ...),
# );

/**
 * String overrides:
 *
 * To override specific strings on your site with or without enabling locale
 * module, add an entry to this list. This functionality allows you to change
 * a small number of your site's default English language interface strings.
 *
 * Remove the leading hash signs to enable.
 */
# $conf['locale_custom_strings_en'] = array(
#   'forum'      => 'Discussion board',
#   '@count min' => '@count minutes',
# );


lib/Drupal/Credentials.pm
package Drupal::Credentials;

use warnings;
use strict;

#use URI::Escape; or regex?

=head1 NAME

Drupal::Credentials - Access credentials (user, pass etc.) from
Drupal's sites directory (settings.php)

=head1 VERSION

Version 0.02

=cut

our $VERSION  = '0.02';
$Drupal::Credentials::Symlinks ='dontfollow';    # default is dontfollow or set to follow

=head1 SYNOPSIS

Access info from Drupal's $db_uri contained in settings.php

    use Drupal::Credentials;

#new
	Drupal::Credentials::Symlinks='follow';#default is dontfollow

    my $credentials = Drupal::Credentials->new($site_dir);

	foreach my $site_id ($credentials->list) {
		my $dbstring=$credentials->get_dbstring($site_id);
		my $proto=$credentials->get_scheme($site_id);
		my $passw=$credentials->get_pass($site_id);
		my $db=$credentials->get_database($site_id);
		my $host=$credentials->get_host($site_id);
		my $user=$credentials->get_user($site_id);
#new:
		my $dir=$credentials->get_install_dir ($site_id);
	}
	my $sites_dir=$credentials->get_sites_dir;


=head1 METHODS

=head2 my $credentials = Drupal::Credentials->new($site_dir);

$site_dir is the full path to the drupal/sites directory.

Returns the Drupal::Credentials object on success and nothing on failure.

=cut

sub new {

	my $class     = shift;
	my $sites_dir = shift;
	return if ( !$sites_dir );
	return if ( !-e $sites_dir );

	my $self = {};
	$self->{sites_dir} = $sites_dir;

	bless( $self, $class );
	$self->_parse_sites_dir;
	return $self;
}

=head2 $credentials->_parse_sites_dir

Reads the directory specified as sites_dir during new, and looks for
subdirectories which a setting.php file in them. Settings.php file
is searched for $db_url from which the credentials are extracted.

TODO: Currently, always returns 1. Should be able to fail

=cut

sub _parse_sites_dir {
	my $self = shift;

	opendir( my $dh, $self->{sites_dir} ) || die "can't opendir sites dir: $!";

	#ignore the 'all' directory
	my @sites = grep { !/^\.|all/ && -d "$self->{sites_dir}/$_" } readdir($dh);
	closedir $dh;

	#weed out symlinks unless Drupal::Credentials::Symlinks = 'follow'
	if ( $Drupal::Credentials::Symlinks eq 'dontfollow' ) {
		print "GET HERE $Drupal::Credentials::Symlinks\n";
		@sites=grep (! -l "$self->{sites_dir}/$_",@sites);
	}

	my @dbstrings;
	foreach my $site_id (@sites) {
		my $path = "$self->{sites_dir}/$site_id";

		#		$self->{sites}{$site_id}='';

		if ( -f "$path/settings.php" ) {
			open( SETTINGS, "<$path/settings.php" );
			$self->{sites}{$site_id}{install_dir}=$path;
			while (<SETTINGS>) {
				if ( $_ =~ /^\s*\$db_url\s*=\s*'(\w+:\/\/[^']+)'/ ) {
					$self->{sites}{$site_id}{dbstring} = $1;
					$self->_parse_dbstring($site_id);
				}
			}
			close SETTINGS;
		}
	}
	return 1;    #indicate success
}

sub _parse_dbstring {
	my $self    = shift;
	my $site_id = shift;

	#Which characters are allowed for the components? Mysql is pretty
	#permissive. It allows at least ASCII if not unicode characters in
	#some settings and cases, but Drupal parses input from $db_uri thru
	#php's parse_url() function and requires URI hex encodings. Quote
	#from settings.php:
	#   : = %3a   / = %2f   @ = %40
	#   + = %2b   ( = %28   ) = %29
	#   ? = %3f   = = %3d   & = %26"

	#perl \w is the word character (alphanumeric or _) [0-9a-zA-Z_]

	$self->{sites}{$site_id}{dbstring} =~ /
	   ([\w\-\%\#]+):\/\/ 		#scheme
	   ([\w\-\%\#]+): 	#user
	   ([\w\-\%\#]+)@ 	#pass
	   ([\w\-\%\#]+)\/  	#host
	   ([\w\-\%\#]+)		#db
	  /x;

	$self->{sites}{$site_id}{scheme} = $1;
	$self->{sites}{$site_id}{user}   = $2;
	$self->{sites}{$site_id}{pass}   = $3;
	$self->{sites}{$site_id}{host}   = $4;
	$self->{sites}{$site_id}{db}     = $5;

	#unescape if uri encoded
	foreach (qw/scheme user pass host db/) {
		$self->{sites}{$site_id}{$_} =~ s/%([0-9A-Fa-f]{2})/chr(hex($1))/eg;

		#print "$self->{sites}{$site_id}{$_}\n";
	}

	#TODO
	#DRUPAL's multiple connections currently not supported!
	#" To specify multiple connections to be used in your site (i.e. for
	# complex custom modules) you can also specify an associative array
	# of $db_url variables with the 'default' element used until otherwise
	# requested."

}

=head2 my $host=$credentials->get_host($site_id);

Return host from credentials.

=cut

#ALTERNATIVE WOULD BE TO USE AUTOLOAD. will not be faster, but
#cleaner code

sub get_host {
	my $self    = shift;
	my $site_id = shift;

	#return empty handed when no site specified
	return if ( !$site_id );
	return $self->{sites}{$site_id}{host};
}

=head2 my $pw=$credentials->get_pass($site_id);

Return pass from credentials for site with $site_id.

=cut

sub get_pass {
	my $self    = shift;
	my $site_id = shift;

	return if ( !$site_id );
	return $self->{sites}{$site_id}{pass};
}

=head2 my $proto=$credentials->get_scheme($site_id);

Return scheme from credentials for site with $site_id.

=cut

sub get_scheme {
	my $self    = shift;
	my $site_id = shift;

	return if ( !$site_id );
	return $self->{sites}{$site_id}{scheme};
}

=head2 my $un=$credentials->get_user($site_id);

Return user from credentials for site with $site_id.

=cut

sub get_user {
	my $self    = shift;
	my $site_id = shift;

	return if ( !$site_id );
	return $self->{sites}{$site_id}{user};
}

=head2 my $db=$credentials->get_database($site_id);

Return database from credentials for site with $site_id.

=cut

sub get_database {
	my $self    = shift;
	my $site_id = shift;

	#return empty handed when no site specified
	return if ( !$site_id );
	return $self->{sites}{$site_id}{db};
}

=head2 my $dbstring=$credentials->get_dbstring($site_id);

Return string from settings.php's $db_uri for site with $site_id. This one
will not be unescaped, if it uses hex encoding (e.g. %2a for :). I suspect
that this method is not of real use except for debugging.

=cut

sub get_dbstring {
	my $self    = shift;
	my $site_id = shift;

	#return empty handed when no site specified
	return if ( !$site_id );
	return $self->{sites}{$site_id}{db};
}

=head2 my $credentials->list;
Return site ids in an array. A site id is the name of the directory in sites
directory, e.g.
	exampledomain.com

(In code site id is sometimes referred to as site name.)

=cut

sub list {
	my $self = shift;
	return keys %{ $self->{sites} };

}

=head2 my $sites_dir=$credentials->get_install_dir;

Return path for the installation (in Drupal's site directory). At the moment,
it can be relative or absolute path.

Todo: Do i need an absolute path?

=cut

sub get_install_dir {
	my $self = shift;
	my $site_id = shift;

	#return empty handed when no site specified
	return if ( !$site_id );
	return $self->{sites}{$site_id}{install_dir};
}



=head2 my $sites_dir=$credentials->get_sites_dir;

Return Drupal's path for site as specified during new.

=cut

sub get_sites_dir {
	my $self = shift;
	return $self->{sites_dir};
}

=head1 SEE ALSO

http://php.net/manual/en/function.parse-url.php
Drupal's settings.php

=head1 AUTHOR

Maurice Mengel, C<< <mauricemengel at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-drupal-credentials at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Drupal-Credentials>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Drupal::Credentials


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Drupal-Credentials>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Drupal-Credentials>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Drupal-Credentials>

=item * Search CPAN

L<http://search.cpan.org/dist/Drupal-Credentials/>

=back

=head1 INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Maurice Mengel.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;    # End of Drupal::Credentials
